src/btree/row_srch.c

```diff
@@ -102,7 +105,22 @@ __wt_search_insert(
     match = skiphigh = skiplow = 0;
     ins = last_ins = NULL;
     for (i = WT_SKIP_MAXDEPTH - 1, insp = &ins_head->head[i]; i >= 0;) {
-        if ((ins = *insp) == NULL) {
+        /*
+         * The ins value within the loop should not change. While the compiler can change the code
+         * in a way that it reads the ins value from memory again in the following code thus
+         * breaking this assumption. In addition, this function also assumes that if a new key is
+         * inserted to the skip list concurrently and we have seen it in the upper stack of the skip
+         * list, we must also see it in the lower stack. Otherwise, the prefix compare optimization
+         * will break if we skip the newly inserted key in the lower stack if we have compared with
+         * it in the upper stack (please refer to WT-10461 for details). Though the insertion of the
+         * new key to the skip list is ordered from lower stack to upper stack, in the weakly
+         * ordered architectures, such as ARM, it is still possible to read an old value and skip
+         * the newly inserted key if the CPU reorders the read.
+         *
+         * Place a read barrier here to avoid these issues.
+         */
+        WT_ORDERED_READ64_LIGHTWEIGHT(&ins, insp);
```

Comment:
Is this new macro just a convenience for this fix, or is it a proposal for how WT should be handling these things in the future.

I think it will be challenging for programmers to understand where and when they need `WT_ORDERED_READ()` compared to `WT_ORDERED_READXX_LIGHTWEIGHT()`. I can live with this as a short term fix. But in the long run I think we should aim for programmers to just express their requirements or expectations and have the underlying primitives do whatever is right for the hardware platform. For example, can we come up with a smarter implementation of `WT_ORDERED_READ()` that we can use everywhere so the programmer can just state where they need to ensuring ordering in the face or concurrency and have the underlying implementation ensure that we do that in the best way (or at least a reasonable way) for the underlying platform?

Reply:
I don't think providing a single implementation is possible. There are actually two different cases.
* We need a read barrier both in x86 and arm
* We only need a weaker barrier in arm

I don't think we can smartly figure out which case it is in the underlying implementation. If we have to pass a parameter to tell the macro what is the correct thing to do, it is the same as introducing a new macro.

In addition, the arm weaker fence is an atomic load. It only works on a memory read, while WT_ORDERED_READ works on general reads.

I think we should know which type of barrier and where to put them when we write the code.

Reply:
I am also concerned about this. We haven't used acquire/release semantics on individual reads/writes until now (other than when defaulted by CAS operations). We rely on explicit memory barriers (read, write or full) whenever needed. This is a digression from that model. Why don't we use a WT_ORDERED_READ here and test if there is any regression on stronger models like x86_64.

Reply:
I prefer this way as well. @agorrod prefer we should not introduce unnecessary barriers to x86.

Reply:
Ok, that would be a good idea if we discussed it, probably offline. Usually, a load acquire goes along with a store release. The platform might prevent the re-order of only the concerned memory location and not the ones accessible through that location. Last year we went through an exercise of looking into details on these and various ARM implementations. We realised that GCC implementations were not always correct, and it is safer to use barrier instructions instead as they correlate with the memory model we have been following until now. Let's talk offline. Sorry if this has been discussed before and I am only getting to the speed now.

Reply:
My preference is that we identify this as a different type of required synchronisation. I'm less concerned about the efficiency aspect, and more concerned about clarity. Specifically, I don't want someone coming along in the future and removing the read barrier because they assume the atomic_cas on the write path is adequate.

To that end I'm not a fan of the `LIGHTWEIGHT` word. I like Keith's suggestion of having an option for the existing WT_ORDERED_READ macro that indicates what problem it is solving.
OTOH: This is different to that. In that the ordered read issues a generic barrier, so doesn't need to be typed. Wheras this wraps the assignment in an atomic so needs to be type aware.

Reply:
To be clear I care about this in the following order of priority:
* Correct code
* Obviously correct
* Performance

What I'm trying to push for is something that makes the code more obviously correct (since it certainly isn't obvious, I think we need something explicit to make it more so)

Reply:
I just wanted to say I'm totally OK with using `WT_ORDERED_READ` here. I still think we should move to a C11 atomics model eventually, but I think it makes sense to use the current primitives in this fix. That this is more expensive than it needs to for x86 is part of what makes moving to C11 atomics desirable IMHO.

> Last year we went through an exercise of looking into details on these and various ARM implementations. We realised that GCC implementations were not always correct
@sulabhM I'm interested in this, can you share more details?

Reply:
@dgomezferro: A lot of that investigation happened in [WT-8959](https://jira.mongodb.org/browse/WT-8959).